File: app/src/main/java/com/dumuzeyn/mp3player/QueueRemovalPlan.java
package com.dumuzeyn.mp3player;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;

/** Deterministic current-item selection after removing library tracks from a queue. */
public final class QueueRemovalPlan {
    public final ArrayList<String> remaining;
    public final int currentIndex;
    public final boolean currentRemoved;

    private QueueRemovalPlan(ArrayList<String> remaining, int currentIndex,
            boolean currentRemoved) {
        this.remaining = remaining;
        this.currentIndex = currentIndex;
        this.currentRemoved = currentRemoved;
    }

    public static QueueRemovalPlan create(List<String> queue, int currentIndex,
            Set<String> removedIds) {
        ArrayList<String> remaining = new ArrayList<>();
        int retainedBeforeCurrent = 0;
        String current = currentIndex >= 0 && currentIndex < queue.size()
                ? queue.get(currentIndex) : "";
        for (int index = 0; index < queue.size(); index++) {
            String id = queue.get(index);
            if (!removedIds.contains(id)) {
                remaining.add(id);
                if (index < currentIndex) {
                    retainedBeforeCurrent++;
                }
            }
        }
        boolean currentRemoved = removedIds.contains(current);
        int nextIndex = remaining.isEmpty() ? 0
                : Math.min(retainedBeforeCurrent, remaining.size() - 1);
        return new QueueRemovalPlan(remaining, nextIndex, currentRemoved);
    }
}

File: app/src/main/java/com/dumuzeyn/mp3player/QueueRemovalPlan.kt
package com.dumuzeyn.mp3player

/** Deterministic current-item selection after removing library tracks from a queue. */
data class QueueRemovalPlan(
    @JvmField val remaining: ArrayList<String>,
    @JvmField val currentIndex: Int,
    @JvmField val currentRemoved: Boolean,
) {
    companion object {
        @JvmStatic
        fun create(queue: List<String>, currentIndex: Int, removedIds: Set<String>): QueueRemovalPlan {
            val remaining = ArrayList<String>(queue.size)
            var retainedBeforeCurrent = 0
            val current = queue.getOrNull(currentIndex).orEmpty()
            queue.forEachIndexed { index, id ->
                if (id !in removedIds) {
                    remaining += id
                    if (index < currentIndex) retainedBeforeCurrent++
                }
            }
            val nextIndex = if (remaining.isEmpty()) 0 else retainedBeforeCurrent.coerceAtMost(remaining.lastIndex)
            return QueueRemovalPlan(remaining, nextIndex, current in removedIds)
        }
    }
}
